python/ray/serve/api.py

```diff
@@ -253,78 +254,79 @@ def deployment(func_or_class: Callable) -> Deployment:
 
 @overload
 def deployment(
-    name: Optional[str] = None,
-    version: Optional[str] = None,
-    num_replicas: Optional[int] = None,
-    init_args: Optional[Tuple[Any]] = None,
-    init_kwargs: Optional[Dict[Any, Any]] = None,
-    route_prefix: Union[str, None, DEFAULT] = DEFAULT.VALUE,
-    ray_actor_options: Optional[Dict] = None,
-    user_config: Optional[Any] = None,
-    max_concurrent_queries: Optional[int] = None,
-    autoscaling_config: Optional[Union[Dict, AutoscalingConfig]] = None,
-    graceful_shutdown_wait_loop_s: Optional[float] = None,
-    graceful_shutdown_timeout_s: Optional[float] = None,
-    health_check_period_s: Optional[float] = None,
-    health_check_timeout_s: Optional[float] = None,
+    name: Default[str] = DEFAULT.VALUE,
+    version: Default[str] = DEFAULT.VALUE,
+    num_replicas: Default[int] = DEFAULT.VALUE,
+    init_args: Default[Tuple[Any]] = DEFAULT.VALUE,
+    init_kwargs: Default[Dict[Any, Any]] = DEFAULT.VALUE,
+    route_prefix: Default[Union[str, None]] = DEFAULT.VALUE,
+    ray_actor_options: Default[Dict] = DEFAULT.VALUE,
+    user_config: Default[Any] = DEFAULT.VALUE,
+    max_concurrent_queries: Default[int] = DEFAULT.VALUE,
+    autoscaling_config: Default[Union[Dict, AutoscalingConfig]] = DEFAULT.VALUE,
+    graceful_shutdown_wait_loop_s: Default[float] = DEFAULT.VALUE,
+    graceful_shutdown_timeout_s: Default[float] = DEFAULT.VALUE,
+    health_check_period_s: Default[float] = DEFAULT.VALUE,
+    health_check_timeout_s: Default[float] = DEFAULT.VALUE,
 ) -> Callable[[Callable], Deployment]:
     pass
 
 
 @PublicAPI(stability="beta")
 def deployment(
     _func_or_class: Optional[Callable] = None,
-    name: Optional[str] = None,
-    version: Optional[str] = None,
-    num_replicas: Optional[int] = None,
-    init_args: Optional[Tuple[Any]] = None,
-    init_kwargs: Optional[Dict[Any, Any]] = None,
-    route_prefix: Union[str, None, DEFAULT] = DEFAULT.VALUE,
-    ray_actor_options: Optional[Dict] = None,
-    user_config: Optional[Any] = None,
-    max_concurrent_queries: Optional[int] = None,
-    autoscaling_config: Optional[Union[Dict, AutoscalingConfig]] = None,
-    graceful_shutdown_wait_loop_s: Optional[float] = None,
-    graceful_shutdown_timeout_s: Optional[float] = None,
-    health_check_period_s: Optional[float] = None,
-    health_check_timeout_s: Optional[float] = None,
+    name: Default[str] = DEFAULT.VALUE,
+    version: Default[str] = DEFAULT.VALUE,
+    num_replicas: Default[Optional[int]] = DEFAULT.VALUE,
+    init_args: Default[Tuple[Any]] = DEFAULT.VALUE,
+    init_kwargs: Default[Dict[Any, Any]] = DEFAULT.VALUE,
+    route_prefix: Default[Union[str, None]] = DEFAULT.VALUE,
+    ray_actor_options: Default[Dict] = DEFAULT.VALUE,
+    user_config: Default[Optional[Any]] = DEFAULT.VALUE,
```

Comment:
Please correct me if i am wrong, we are having `DEFAULT.VALUE` is `1`,  what does it mean for setting every attribute to `1` here?

Reply:
It’s not quite setting every attribute to 1 here. `DEFAULT.VALUE` is a [Python enum element](https://docs.python.org/3/library/enum.html#creating-an-enum) with an arbitrary value (in this case 1). It’s not an alias for 1. It could be set to any other arbitrary value (e.g. 2, “hello”, etc.) and the code here would still be the same.

I’m using `DEFAULT.VALUE` as the default because the user will never pass that in as a value for a deployment option (since it’s meaningless to the user and it would require them to import the enum from the Serve codebase). So in the decorator body, if any attributes are set to `DEFAULT.VALUE`, we know that the user didn’t set them. That lets us populate `user_configured_options` correctly.

Reply:
```
In [3]: from enum import Enum, auto

In [4]: class DEFAULT(Enum):
   ...:     VALUE = auto()
   ...:

In [5]: DEFAULT.VALUE
Out[5]: <DEFAULT.VALUE: 1>

In [6]: DEFAULT.VALUE == 1
Out[6]: False

In [7]: DEFAULT.VALUE == DEFAULT.VALUE
Out[7]: True
```